src/hitrecord.nim

```diff
@@ -1,276 +1,164 @@
 
 
-type
-    HitRecord* = object
-        ray*: Ray
-        t*: float32
-        shape*: ShapeKind
-        material*: Material
-
-        world_pt*: Point3D
-        surface_pt*: Point2D
-        normal*: Normal
+proc checkIntersection*(aabb: Interval[Point3D], ray: Ray): bool {.inline.} =
+    # this is done in the same reference system
+    let 
+        (min, max) = (aabb.min - ray.origin, aabb.max - ray.origin)
+        txSpan = newInterval(min.x / ray.dir[0], max.x / ray.dir[0])
+        tySpan = newInterval(min.y / ray.dir[1], max.y / ray.dir[1])
 
+    if txSpan.min > tySpan.max or tySpan.min > txSpan.max: return false
 
-proc allHitTimes*(shape: Shape, ray: Ray): Option[seq[float32]] =
-    var t_hit: float32
-    let inv_ray = ray.transform(shape.transform.inverse)
+    let tzSpan = newInterval(min.z / ray.dir[2], max.z / ray.dir[2])
+    
+    var hitSpan = newInterval(max(txSpan.min, tySpan.min), min(txSpan.max, tySpan.max))
+    if hitSpan.min > tzSpan.max or tzSpan.min > hitSpan.max: return false
 
-    case shape.kind
-    of skTriangle: discard
+    return true
 
-    of skTriangularMesh: discard            
 
-    of skAABox: discard
+proc getHitLeafs*(subScene: SubScene; ray: Ray): Option[seq[SceneNode]] =
+    if subScene.tree.isNil: return none seq[SceneNode]
+    if not checkIntersection(subScene.tree.aabb, ray): return none seq[SceneNode]
 
-    of skSphere:
-        let (a, b, c) = (norm2(inv_ray.dir), dot(inv_ray.origin.Vec3f, inv_ray.dir), norm2(inv_ray.origin.Vec3f) - 1)
-        let delta_4 = b * b - a * c
-        if delta_4 < 0: return none(seq[float32])
+    var sceneNodes: seq[SceneNode]
+    case subScene.tree.kind
+    of nkLeaf: 
+        sceneNodes.add subScene.tree
+    of nkBranch:
+        for node in [subScene.tree.left, subScene.tree.right]:
+            if node != nil:
+                let hits = getHitLeafs((subScene.rs, node), ray)
+                if hits.isSome: sceneNodes = concat(sceneNodes, hits.get)
 
-        let (t_l, t_r) = ((-b - sqrt(delta_4)) / a, (-b + sqrt(delta_4)) / a)
-        if t_l > ray.tmin and t_l < ray.tmax and t_r > ray.tmin and t_r < ray.tmax: 
-            return some(@[t_l, t_r])
-        elif t_l > ray.tmin and t_l < ray.tmax:
-            return some(@[t_l])
-        elif t_r > ray.tmin and t_r < ray.tmax:
-            return some(@[t_r])
-        return none(seq[float32])
+        if sceneNodes.len == 0: return none seq[SceneNode]
 
-    of skPlane:
-        if abs(inv_ray.dir[2]) < epsilon(float32): return none(seq[float32])
-        t_hit = -inv_ray.origin.z / inv_ray.dir[2]
-        if t_hit < ray.tmin or t_hit > ray.tmax: return none(seq[float32])
+    some sceneNodes
 
-        return some(@[t_hit, Inf])
 
-    of skCylinder: discard
+type HitPayload* = object
+    shape*: Shape
+    ray*: Ray
+    t*: float32
+    
+proc newHitPayload(handler: ShapeHandler, ray: Ray): Option[HitPayload] =
+    let invRay = ray.transform(handler.transformation.inverse) 
```

Comment:
Do you think that it could be a good idea to make code less verbose? Sometimes I feel like we could just avoid too write full names